File: 015/problem015.hs
memo = replicate 20 (replicate 20 [])

pathCount :: Int -> Int -> Int -> Int -> Integer
pathCount r c row_max col_max = 
    if r == row_max && c == col_max then
        1
    else if r < row_max && c < col_max sthen
        (pathCount (r + 1) c row_max col_max) + (pathCount r (c + 1) row_max col_max)
    else if r < row_max then
        pathCount (r + 1) c row_max col_max
    else
        pathCount r (c + 1) row_max col_max

main = do
    print (pathCount 0 0 20 20)

File: 015/problem015.c
#include <stdio.h>

long pathCount(int r, int c, int row_max, int col_max, long memo[row_max + 1][col_max + 1]) {
    long count = 0;
        if (r < row_max) {
            if (memo[r + 1][c]) {
                count += memo[r + 1][c];     
            } else {
                count += pathCount(r + 1, c, row_max, col_max, memo);
            }
        }
        if (c < col_max) {
            if (memo[r][c+1]) {
                count += memo[r][c + 1];
            } else {
                count += pathCount(r, c + 1, row_max, col_max, memo);
            }
        }
        memo[r][c] = count;
        if (r == row_max && c == col_max) {
            count++;
            memo[r][c] == 1;
        }
        return count;
}


int main(int argc, char *argv[]) {

    long memo[21][21] = { {0} };
    printf("%ld\n", pathCount(0, 0, 20, 20, memo));
    return 0;
}
